design-doc.md

```diff
@@ -0,0 +1,128 @@
+
+As for the end of the listener, it is possible to request the listener to terminate automatically after certain number of entries have been scanned or after certain order key has been reached. On top of that, the listener itself can choose to stop itself when certain criteria has been met. Since listeners are independent, the termination criteria does not have to be consistent, even within the very same listener.
+
+Implementation-wise, listening to a stream spawns a dedicated thread. (`TODO(dkorolev): The design enforces the users to be "good citizens", and not spawn listeners in large quantities, however, in certain cases one thread per listener might not be the most efficient approach. I should document it in a bit more detail.`).
+
+Being run in dedicated threads, listeners never interfere with each other and with the publisher. It is legit for a very slow listener to run in parallel with the one that is expected to always be caught up with the most recently added entires.
+
+A per-entry handler of a listener has the means to tell how far "behind" on a stream it is, both index-wise and order-key-wise. This is important for serving guarantees: for example, when a listener is responsible for serving external requests, the common design is to notify its master or load balancer that it is healthy as it is caught up and that it still is in the "replay log" mode.
+
+### Publishing
+
+While publishing, entries should come in increasing order of order keys.
+
+By design, a `Publish()` operation is not thread-safe, and it is advised to acquire stream's publisher within a specific thread that does the publishing.
+
+Implementation-wise, within KnowSheet, this thread would almost always be an HTTP server, a consumer of an in-memory message queue, a listener of another stream or a producer of a TailProduce job. All of the above are single-threaded by design, and they respect the order of calls.
+
+Streams are hardly deleted: in most production systems, if a stream is created it persists for the lifetime of the binary.
+
+With that in mind, however, it is possible to shutdown a stream without terminating the binary. The shutdown operation is designed to be as safe as possible: implementation-wise, it will wait for all pending calls to currently active listeners, as well as signal the acquired publisher to terminate. Only after all the handles have been released, the stream will be destructed.
+
+
+## Replication
+
+From data integrity standpoint, each stream is assigned an **authority**: A binary, running, usually a dedicated machine, that is responsible to reporting the current up-to-date status of the stream.
```

Comment:
1) **a** binary (lowercase "a")
2) "responsible **for reporting**" or "responsible **to report**"